src/core/SpriteAnimator.tsx

```diff
@@ -1,72 +1,176 @@
 import * as React from 'react'
-import { useFrame, Vector3, MeshProps } from '@react-three/fiber'
+import { useFrame, Vector3 } from '@react-three/fiber'
 import * as THREE from 'three'
 import { Instances, Instance } from './Instances'
+import { Billboard } from './Billboard'
 import { useSpriteLoader } from './useSpriteLoader'
+import { SetStateAction } from 'react'
+
+// Frame-related types
+type SourceSize = {
+  w: number
+  h: number
+}
+
+type FrameData = {
+  frame: {
+    x: number
+    y: number
+  }
+  sourceSize: SourceSize
+}
+
+type MetaData = {
+  version: string
+  size: {
+    w: number
+    h: number
+  }
+  scale: string
+}
+
+type SpriteData = {
+  frames: Record<string, FrameData[]> | FrameData[]
+  meta: MetaData
+}
+
+type FrameBuffer = {
+  key: number
+  frames: Record<string, FrameData[]> | FrameData[]
+  selectedFrame: string
+  offset: {
+    x: number
+    y: number
+  }
+}
+
+type AnimationEventData = {
+  currentFrameName: string
+  currentFrame: number
+}
 
 type CommonMeshProps = {
   frustumCulled?: boolean
   renderOrder?: number
   visible?: boolean
   userData?: any
+  layers?: THREE.Layers
   matrixAutoUpdate?: boolean
 }
 
 export type SpriteAnimatorProps = {
+  /** The start frame of the animation */
   startFrame?: number
+  /** The end frame of the animation */
   endFrame?: number
+  /** The desired frames per second of the animation. If set to 0 or negative, animation will be static */
   fps?: number
+  /** The frame identifier to use, must be one of animationNames */
   frameName?: string
+  /** The URL of the texture JSON (if using JSON-Array or JSON-Hash) */
   textureDataURL?: string
+  /** The URL of the texture image */
   textureImageURL?: string
+  /** Whether or not the animation should loop */
   loop?: boolean
+  /** The number of frames of the animation (required if using plain spritesheet without JSON) */
   numberOfFrames?: number
+  /** Whether or not the animation should auto-start when all assets are loaded */
   autoPlay?: boolean
+  /** The animation names of the spritesheet (if the spritesheet -with JSON- contains more animation sequences) */
   animationNames?: Array<string>
-  onStart?: Function
-  onEnd?: Function
-  onLoopEnd?: Function
-  onFrame?: Function
+  /** Event callback when the animation starts or restarts
+   * @param data Animation state data containing currentFrameName and currentFrame
```

Comment:
not sure `@param` is really useful is since it is already typed, ie: `: AnimationEventData`